### ISSUE TITLE
memory leak in tf.py_function

### ISSUE THREAD

**Issue by sjtusmartboy** [tensorflow/tensorflow#51839]
tensorflow 2.7 python 3.7

minimum code

```
import tensorflow as tf
import numpy as np
import os

def test(xyz_batch,  k):
    indices = np.zeros((1000, 1000, 20), dtype=np.int64)
    dist = np.zeros((1000, 1000, 20), dtype=np.float32)
    return indices, dist


while True:
    ret = tf.py_function(test, [0,0], [tf.int64, tf.int64])
```

**Comment by sjtusmartboy** [comment]
same memory leak issue
```
def test(indices,  dist):
    # indices = np.zeros((1000, 1000, 20), dtype=np.int64)
    # dist = np.zeros((1000, 1000, 20), dtype=np.int64)
    return indices, dist


while True:
    indices = np.zeros((1000, 1000, 20), dtype=np.int64)
    dist = np.zeros((1000, 1000, 20), dtype=np.int64)
    ret = tf.py_function(test, [indices,dist], [tf.int64, tf.int64])
```

**Comment by sjtusmartboy** [comment]
with the below as the comparison of no memory leak
```
def test(indices,  dist):
    # indices = np.zeros((1000, 1000, 20), dtype=np.int64)
    # dist = np.zeros((1000, 1000, 20), dtype=np.int64)
    print(1)
    return indices, dist


while True:
    indices = np.zeros((1000, 1000, 20), dtype=np.int64)
    dist = np.zeros((1000, 1000, 20), dtype=np.int64)
```

**Comment by sushreebarsa** [comment]
@Saduf2019 Was able to reproduce the issue on Colab using TF 2.6 and tf-nightly,Please find the [gist](https://colab.research.google.com/gist/sushreebarsa/a739fc42e9dcd921c83591c773565f0a/untitled434.ipynb) here for reference.Thank you!

**Comment by MrinalTyagi** [comment]
@sushreebarsa  @Saduf2019 I would like to work on this issue. Could you guide me with the same?

**Comment by sjtusmartboy** [comment]
@sushreebarsa @Saduf2019 Please make some updates, even this issue is very hard, thanks!

**Comment by sjtusmartboy** [comment]
@sushreebarsa @Saduf2019 @ymodak Please take this issue seriously because the memory issue is intolerant in any product. I still remember in my university class, my C++ teacher gave the low mark in the final exam because of forgetting to free the memory. I would also say that if I do not treat the memory issue seriously in the company, my boss would fire me. So I hope tensorflow would and should be a successful product, ever and forever. It cannot do so without your diligence. So make it and we believe it.

**Comment by rainwoodman** [comment]
Hi,

Thanks for the investigation.

This looks like a known issue, https://github.com/tensorflow/tensorflow/issues/35084 which was closed due to inactivity.

The error is caused by py_function creating a new entry in tape_cache on every call, where it stores a new reference to the input arguments and the function for tape. The tape_cache entry is consumed (via popping) during gradient calculation.

This is the necessary overhead for recording the tape. The problem is we are incurring it when there is no active Tape.
I haven't thought through how best to fix this -- maybe we can detect whether there is an active tape and act accordingly.

Meanwhile, here are a few workarounds that I can think of:

- (mentioned in the docstring of _eager_py_func) use the internal function _eager_py_func, and set use_tape_cache=False. This disables gradient support I believe (but I haven't tried).
- consume the tape_cache entry by computing the gradient per loop entry. This is the typical pattern in a training loop.
- (as you have shown in the previous comment) use numpy_function, which plays nicely with tapes.

**Comment by sjtusmartboy** [comment]
@rainwoodman Thanks for your professional reply. I've tried rewriting all the numpy code to tf code, but there's still memory leak. I do not know how to find the bugs not, currently no any method to analyze. Is there any function like graph.finalize() in tf 1.x?

**Comment by rainwoodman** [comment]
Sorry I do not know much about graph.finalize() or tf 1.x to directly answer your question.

On detecting memory leak, the test case for #35084  may serve as an example:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/data/kernel_tests/memory_cleanup_test.py#L151

FYI: here is the workaround used by #35084, via _eager_py_func.
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/data/ops/dataset_ops.py#L1082

The bug manifests as memory leak when not requesting gradients during eager mode execution of EagerPyFunc (via py_function) and also as memory leak during graph execution (py_function inside tf.function), so I think you will be affected it as long as you use tf.py_function and you do not use gradients.

We are working on a tentative fix, there is a possibility the fix makes into the 2.7 release.

PS: In case you are boarding tensorflow recently, I strongly recommend reading the [intro_to_graphs guide](https://www.tensorflow.org/guide/intro_to_graphs), and other guides on the doc site to get the most out of tensorflow 2, if you haven't done so.

**Comment by sjtusmartboy** [comment]
@rainwoodman Thanks again very much for helping me find the solutions. I will try the memory cleanup test. Here I find a fix relating to the memory leak,
https://stackoverflow.com/questions/58915522/tf2-0-memory-leak-from-applying-keras-model-to-symbolic-tensor

I wonder whether this issue has being fixed? Is it a must to put the loop inside the tf.function?

**Comment by rainwoodman** [comment]
Hmm... I am not aware of a "Keras + random in a loop" leak or if there is a fix (but I am also new to the TF codebase).

If your use case is affected, what about filing a new issue with a reproducer? We can continue from there.

**Comment by sjtusmartboy** [comment]
@rainwoodman That's just the solution! The loop must be placed inside the tf.function, otherwise there will be the memory leak. Hope someone will work on this issue,after all this issue has been on post for over a year.

**Comment by sjtusmartboy** [comment]
Please do not close this issue as it involves two bugs. The first is the tf.py_function and the second is that the loop should be inside the tf.function referenced in https://stackoverflow.com/questions/58915522/tf2-0-memory-leak-from-applying-keras-model-to-symbolic-tensor.So please reopen it.

**Comment by rainwoodman** [comment]
@sjtusmartboy:

The SO example runs in eager mode, calls tf.random, and doesn't call tf.py_function. 

Did you encounter the leak in a similar context?

**Comment by sachinprasadhs** [comment]
@sjtusmartboy Could you please update on the above request.

**Comment by TheBigTicket13** [comment]
I have encounter this issue recently. Solution for me was to avoid tf.py_function. I used normal python function instead with eager execution enabled. In this mode it is possible to convert Tensor to numpy with numpy() method

My py_function was inside tf.GradientTape() and yet there was a memory leak

**Comment by ya-b** [comment]
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/script_ops.py

```
  # Store a reference to the function in the graph to ensure it stays alive
  # as long as the graph lives. When the graph is destroyed, the function
  # is left to the garbage collector for destruction as well.
  graph._py_funcs_used_in_graph.append(func)  # pylint: disable=protected-access
 ```
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/framework/ops.py
```
class _DefaultGraphStack(stack.DefaultStack):  # pylint: disable=protected-access
  """A thread-local stack of objects for providing an implicit default graph."""
```
clean the graph may help
